### PR TITLE
OCPBUGS-1577: On-prem: Ensure resolv-prepender respects NM dispatcher timeout

### PR DESCRIPTION
**Pull request by martinkennelly**
~~Enabling debug mode for testing. It will probably generate too much noise in production.~~

### CODE REVIEW

**Comment by martinkennelly** [comment]
Forced pushed to remove debug of NM prepender script. If we left it there it would cause too much noise to logs IMO :(

**Comment by martinkennelly** [comment]
@andreaskaris @jcaamano PTAL

**Comment by jcaamano** [comment]
@martinkennelly can you explain what's going to happen here? Will ' /usr/bin/podman run' just wait until `/usr/bin/podman pull finishes' or will it fail?

**Comment by jcaamano** [comment]
> @martinkennelly can you explain what's going to happen here? Will `/usr/bin/podman run` just wait until `/usr/bin/podman pull` finishes or will it fail?

OK, hopefully `pod image exists` behaves as we would expect from it.

I was thinking though we would not like the script to take more that 30s overall (accounting for up to 3 probable events to be handled related to device activation: up, dhcp4, dhcp6 against the 90s timeout). Aggregating timeouts probably is not the easiest way to go about this, so perhaps doing this inside a function would make more sense
```
function resolv_prepender {
 // all logic here
}

{{if .Proxy -}}
{{if .Proxy.HTTPProxy -}}
export HTTP_PROXY={{.Proxy.HTTPProxy}}
{{end -}}
{{if .Proxy.HTTPSProxy -}}
export HTTPS_PROXY={{.Proxy.HTTPSProxy}}
{{end -}}
{{if .Proxy.NoProxy -}}
export NO_PROXY={{.Proxy.NoProxy}}
{{end -}}
{{end -}}

export IFACE STATUS DHCP6_FQDN_FQDN IP4_DOMAINS IP6_DOMAINS
export -f resolv_prepender
timeout 90s bash -c resolv_prepender
// handle timeout
```

**Comment by martinkennelly** [comment]
> @martinkennelly can you explain what's going to happen here? Will `/usr/bin/podman run` just wait until `/usr/bin/podman pull` finishes or will it fail?

Originally, I saw `podman run` would wait and follow the download of an existing `podman pull` but I just checked there again and it doesnt do that. Its not documented so I put in a retry until image is available or else exit instead.

Since there is no retry for NM dispatcher, this could still fail but its an improvement.

This change need retesting again for multiple default gateways. I will update this PR when its complete.

**Comment by martinkennelly** [comment]
> > @martinkennelly can you explain what's going to happen here? Will `/usr/bin/podman run` just wait until `/usr/bin/podman pull` finishes or will it fail?
> 
> OK, hopefully `pod image exists` behaves as we would expect from it.
> 
> I was thinking though we would not like the script to take more that 30s overall (accounting for up to 3 probable events to be handled related to device activation: up, dhcp4, dhcp6 against the 90s timeout). Aggregating timeouts probably is not the easiest way to go about this, so perhaps doing this inside a function would make more sense
> 
> ```
> function resolv_prepender {
>  // all logic here
> }
> 
> {{if .Proxy -}}
> {{if .Proxy.HTTPProxy -}}
> export HTTP_PROXY={{.Proxy.HTTPProxy}}
> {{end -}}
> {{if .Proxy.HTTPSProxy -}}
> export HTTPS_PROXY={{.Proxy.HTTPSProxy}}
> {{end -}}
> {{if .Proxy.NoProxy -}}
> export NO_PROXY={{.Proxy.NoProxy}}
> {{end -}}
> {{end -}}
> 
> export IFACE STATUS DHCP6_FQDN_FQDN IP4_DOMAINS IP6_DOMAINS
> export -f resolv_prepender
> timeout 30s bash -c resolv_prepender
> // handle timeout
> ```

That seems better :) Will rework.

**Review by mandre — COMMENTED** [review]
Ouch, my bad. Thanks for figuring this out.

**Comment by martinkennelly** [comment]
> OK, hopefully `pod image exists` behaves as we would expect from it.

I talked to a podman expert and he told me:
`By code inspection only, nowadays it’s going to complain about an invalid image and return false`
I asked about "nowadays" and he showed me to commit from June 2021: https://github.com/containers/podman/commit/5fc622f945de46db85c8cc0284f935bac1929e3a
I then checked 4.12 podman version and its 4.2.0 and I see the same changes are in that branch: https://github.com/containers/podman/blob/v4.2.0-rhel/vendor/github.com/containers/common/libimage/runtime.go#L160
This same check must be performed in 4.11.

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml [hunk 1]
```diff
@@ -20,89 +20,140 @@ contents:
...
+            # Image must exist before reaching this point otherwise timeout may cause image corruption
+            NAMESERVER_IP=$(/usr/bin/podman run --rm \
+                --authfile /var/lib/kubelet/config.json \
+                --net=host \
+                {{ .Images.baremetalRuntimeCfgImage }} \
+                node-ip \
+                show \
+                --retry-on-failure \
+                "{{ onPremPlatformAPIServerInternalIP . }}" \
+                "{{ onPremPlatformIngressIP . }}")
```
Comment (mandre) [inline templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:81]: Was this change intentional (you're partially reverting [PR 3269](https://github.com/openshift/machine-config-operator/commit/87a51419c1662177947d6a4b3f5f42cad13d0a66#diff-255f8a4599166f31961853ea8626f969ca4231c55aacbc20a5bb3ceb640f911d))? IMO it should be change it back to:
```suggestion
                {{range onPremPlatformAPIServerInternalIPs . }}"{{.}}" {{end}} \
                {{range onPremPlatformIngressIPs . }}"{{.}}" {{end}})
```

Comment (martinkennelly) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:81]: No. I have no idea how that got removed. Ill review all the changes again.

**Comment by mandre** [comment]
Just a tip for reviewers, it was easier to review the changes when ignoring the whitespaces diff -> https://github.com/openshift/machine-config-operator/pull/3394/files?w=1

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml [hunk 2]
```diff
@@ -20,89 +20,116 @@ contents:
...
+            # image corruption
+            # https://github.com/containers/podman/issues/14003
+            if ! /usr/bin/podman image exists {{ .Images.baremetalRuntimeCfgImage }}; then
+                # Multiple podman pulls may occur in parallel wasting bandwidth therefore this
+                # needs to be removed when podman image corruption issue is resolved.
+                # Send to background to avoid SIGTERM from timeout which may cause image corruption issue during image pull
+                pull_baremetal_runtime_cfg_image &
+                # Wait until image is available
+                >&2 echo "NM resolv-prepender: Waiting for baremetal runtime cfg image to be available"
+                until /usr/bin/podman image exists {{ .Images.baremetalRuntimeCfgImage }}; do
```
Comment (mandre) [inline templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:66]: What happens if the image pull takes longer than the 30s timeout granted to run the `resolv_prepender` function?

I could be missing something obvious, but it looks to me that the pull will be killed because the parent process dies. Check this small script:
```bash
function resolv_prepender {
        longer_than_timeout &
        while :; do
                sleep 1
        done
        echo unreachable
}

function longer_than_timeout {
        for i in 1 2 3; do
                echo $i
                sleep 1
        done
        echo done
}

export -f resolv_prepender longer_than_timeout
timeout 2s bash -c resolv_prepender
```

```
/tmp
❯ bash ./test.sh
1
2

/tmp
❯ echo $?
124
```
If that's really the case, we shouldn't be bothering with running the pull as a separate command. Or alternatively, we should pull the image before we run the dispatcher script. Is that even possible?

Comment (martinkennelly) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml]: Duh. Thank you @mandre. Fixed it. Appreciate your eyes here.

Comment (martinkennelly) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml]: Ive asked a few folks and no one has come up any ideas how to do this after network manager is up but before dispatcher scripts are invoked. If you have any ideas, please let me know.

Comment (martinkennelly) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml]: Regarding whether this child process will continue after the dispatcher script is run:
```
Dispatcher scripts are run one at a time, but asynchronously from the main NetworkManager process, and will be killed if they run for too long. If your script might take arbitrarily long to complete, you should spawn a child process and have the parent return immediately. 
```
https://networkmanager.dev/docs/api/latest/NetworkManager-dispatcher.html

I tested that the child process will continue following return of the dispatcher script.
We dont return immediately, but if we respect NM timeout, we should be ok.

**Review by mandre — COMMENTED** [review]
This new version LGTM, thanks a lot @martinkennelly.

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml [hunk 3]
```diff
@@ -20,89 +20,121 @@ contents:
...
-    esac
+            ;;
+            *)
+            ;;
+        esac
+    }
+
+    # Pull container image outside of the timeout in a background process to workaround possible
+    # image corruption [1]. As long as the overall script duration is below NM dispatcher timeout, this pull will continue
+    # in the background even after this script concludes [1].
```
Comment (jcaamano) [inline templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:122]: hrrm so if the dispatcher script ends, this will be orphaned, detached from parent and inherited by init or whatever the session manager is, so that it continues to run in the background without NM ever being aware of it. Perhaps adding this extra minor detail in the comment might be useful, but this is really a nit, up to you.

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml [hunk 4]
```diff
@@ -20,89 +20,121 @@ contents:
...
+    # [1] https://github.com/containers/podman/issues/14003
+    if ! /usr/bin/podman image exists {{ .Images.baremetalRuntimeCfgImage }}; then
+        # Multiple image pulls may occur in parallel wasting bandwidth therefore this
+        # needs to be removed when podman image corruption issue is resolved.
+        # Send to background to allow it to run indepedently of dispatcher timeout. Image pull will continue as long as
+        # this script respects NM timeout.
```
Comment (jcaamano) [inline templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:128]: `Image pull will continue as long as this script respects NM timeout` is difficult to make sense of. Maybe add more detail or it needs to be rephrased after the latest change? Are you trying to say that if the script is killed because of it not respecting the NM timeout, the pull will be killed as well?
Sorry for the comment nits, but we are doing something weird so it might pay off to be super clear about it in the comments. Maybe
`As long as this script respects and ends by itself before the NM timeout, the pull will continue in the background without being killed by NM`

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml (hunk 3 above)
Comment (jcaamano) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:122]: `If the script ends by itself before the NM timeout and before this pull is done, this pull will be orphaned from the parent and continue running without NM being aware of it`

**Comment by mandre** [comment]
Don't mind the e2e-openstack failure, the cloud is having issue. We've notified them and I'll retest as soon as it's fixed on their side.

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml (hunk 4 above)
Comment (martinkennelly) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml]: > Are you trying to say that if the script is killed because of it not respecting the NM timeout, the pull will be killed as well?

Yes.
If its not clear, I value your input because this is indeed weird.

**Comment by martinkennelly** [comment]
I have tested the case for two default routes and it works as expected.

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml [hunk 5]
```diff
@@ -20,89 +20,124 @@ contents:
...
+    # timeout.
+    # [1] https://github.com/containers/podman/issues/14003
+    if ! /usr/bin/podman image exists {{ .Images.baremetalRuntimeCfgImage }}; then
+        # Multiple image pulls may occur in parallel wasting bandwidth therefore this
+        # needs to be removed when podman image corruption issue is resolved.
+        pull_baremetal_runtime_cfg_image &
+    fi
+
+    export IFACE STATUS DHCP6_FQDN_FQDN IP4_DOMAINS IP6_DOMAINS
+    export -f resolv_prepender pull_baremetal_runtime_cfg_image
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml]: done

**Comment by jcaamano** [comment]
/lgtm
/cc @cybertron 
PTAL

**Comment by mandre** [comment]
openstack cloud is supposedly fixed, let's see
/test e2e-openstack

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml [hunk 6]
```diff
@@ -20,89 +20,124 @@ contents:
...
 
-        # Pull container image outside of the timeout to workaround possible
-        # image corruption
-        # https://github.com/containers/podman/issues/14003
-        /usr/bin/podman pull --authfile /var/lib/kubelet/config.json {{ .Images.baremetalRuntimeCfgImage }}
+            # This function is subject to a timeout. The timeout may interrupt podman run's image pull if the image is
+            # not available and cause image corruption [1]. Image should be available or being pulled before we reach
+            # this point. Wait until its available.
+            # [1] https://github.com/containers/podman/issues/14003
+            until /usr/bin/podman image exists {{ .Images.baremetalRuntimeCfgImage }}; do
```
Comment (cybertron) [inline templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:55]: This can't be done before we populate resolv.conf on line 62. If I'm reading this right, the script will never succeed as written because the image pull won't work until resolv.conf is populated, and this check prevents us from ever populating resolv.conf.

I think we need to factor the resolv.conf creation out and run it first, and then we can start the image pull in the background and only wait to prepend the local IP.

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml [hunk 7]
```diff
@@ -20,89 +20,116 @@ contents:
...
+                {{ .Images.baremetalRuntimeCfgImage }} \
+                node-ip \
+                show \
+                --retry-on-failure \
+                {{range onPremPlatformAPIServerInternalIPs . }}"{{.}}" {{end}} \
+                {{range onPremPlatformIngressIPs . }}"{{.}}" {{end}})
+            DOMAINS="${IP4_DOMAINS} ${IP6_DOMAINS} {{.DNS.Spec.BaseDomain}}"
+            if [[ -n "$NAMESERVER_IP" ]]; then
+                if systemctl -q is-enabled systemd-resolved; then
+                    >&2 echo "NM resolv-prepender: Setting up systemd-resolved for OKD domain and local IP"
```
[the remark this answers is not in the document]
Comment (cybertron) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:78]: FWIW, this was added by someone else (I think for OKD). I don't know that we test it at all in ci so historically I've tried to leave it alone.

I guess the only concern with this new logic is that systemd-resolved will be restarted every time the script runs. I don't know if that causes any disruption in DNS resolution, but if it does we might want to find some way to avoid that. Maybe we hash the file contents or something so we can check if they changed?

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml [hunk 8]
```diff
@@ -20,89 +20,140 @@ contents:
...
-                if ! grep -q search /etc/resolv.tmp; then
-                    # Make sure we have a search entry
-                    echo "search {{.DNS.Spec.BaseDomain}}" >> /etc/resolv.tmp
                 else
-                    # Make sure cluster domain is first in the search list
-                    sed -i "s/^search \(.*\)/search {{.DNS.Spec.BaseDomain}} \1/" /etc/resolv.tmp
-                    # Remove duplicate cluster domain entries
-                    sed -i "s/\(search {{.DNS.Spec.BaseDomain}}.*\) {{.DNS.Spec.BaseDomain}}\( .*\|$\)/\1\2/" /etc/resolv.tmp
+                    >&2 echo "NM resolv-prepender: Prepending 'nameserver $NAMESERVER_IP' to /etc/resolv.conf (other nameservers from /var/run/NetworkManager/resolv.conf)"
+                    RESOLVCONFTMPPATH="$(mktemp)"
```
[the remark this answers is not in the document]
Comment (cybertron) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:92]: Yikes, how did we not catch that the tmp file name was predictable (I'm pretty sure this is my fault, so that's a rhetorical question...)?

**Review by cybertron — CHANGES_REQUESTED** [review]
I don't think this will work, and the fact that all the on-prem ci is failing suggests I'm right. :-)

I'm good with the overall approach, but I left some comments inline about some changes we'll need for this to work. I'm not too concerned about the resolved case because that's kind of supported on a best-effort basis on our end anyway, but we for sure need to fix the chicken-and-egg problem with the ordering of operations here.

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml (hunk 6 above)
Comment (jcaamano) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:55]: `podman image exists` is a local check I think.

But you are right, I think 
```
# Ensure resolv.conf exists and contains nameservers before we try to run podman
if [[ ! -e /etc/resolv.conf ]] || ! grep -q nameserver /etc/resolv.conf; then
    cp /var/run/NetworkManager/resolv.conf /etc/resolv.conf
fi
```
should be moved into `pull_baremetal_runtime_cfg_image`

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (jcaamano) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:78]: We could write 60-kni.conf only if the contents are different, and compare the file tiemstamp with the service start timestamp.

But it does strike me as weird that the service needs to be restarted, there might be another option to make the service read the config without service interruption.

Another option is to set a null TERM trap handler for the portions of the script we don't want to be interrupted.

[the remark this answers is not in the document]
Comment (jcaamano) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:78]: I have not found a way to force systemd-resolved to read the configuration without a restart, but it does read it on its own a few seconds after the change. The only indication I found on the docs about this is 
```
SIGUSR2
Upon reception of the SIGUSR2 process signal systemd-resolved will flush all caches it maintains. Note that it should normally not be necessary to request this explicitly – except for debugging purposes – as systemd-resolved flushes the caches automatically anyway any time the host's network configuration changes. Sending this signal to systemd-resolved is equivalent to the resolvectl flush-caches command, however the latter is recommended since it operates in a synchronous way.
```

So perhaps we should not worry if we are killed between writing the configuration and restarting systemd, and just keep restarting systemd-resolved only if we actually wrote the configuration.

File: templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (martinkennelly) [reply templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml:121]: Wasnt a big deal imo.

**Comment by martinkennelly** [comment]
> FWIW, this was added by someone else (I think for OKD). I don't know that we test it at all in ci so historically I've tried to leave it alone.

I guess the only concern with this new logic is that systemd-resolved will be restarted every time the script runs. I don't know if that causes any disruption in DNS resolution, but if it does we might want to find some way to avoid that. Maybe we hash the file contents or something so we can check if they changed?

@cybertron I added logic to detect if we need to restarted systemd or not.

**Comment by martinkennelly** [comment]
/retest

CI looking good:
e2e-metal-ipi failed because gather extra failed. tests ok.

e2e-metal-ipi-ovn-dualstack:
Reoccurring issue with ClusterOperatorDegraded. Asked Jamo about this. Will file a bug. Unrelated to this PR.
https://search.ci.openshift.org/?search=alert+ClusterOperatorDegraded+fired+for+%5Cd%2B+seconds+with+labels%3A+%5C%7Bname%3D%22network%22%2C+namespace%3D%22openshift-cluster-version%22%2C+reason%3D%22ApplyOperatorConfig%22%2C+severity%3D%22warning%22%5C%7D&maxAge=168h&context=1&type=bug%2Bissue%2Bjunit&name=&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job

e2e-vsphere:
Lots of on-going issues with the storage driver:

https://search.ci.openshift.org/?search=+%5C%5BDriver%3A+vsphere%5C%5D+%5C%5BTestpattern%3A+Generic+Ephemeral-volume+&maxAge=168h&context=1&type=bug%2Bissue%2Bjunit&name=&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job

**Comment by cybertron** [comment]
LGTM pending the remaining ci jobs. Based on what has already passed I expect this will be fine though.

**Comment by martinkennelly** [comment]
Thanks @cybertron for your eyes.
Will remove WIP when Ramon M confirms this fixes the issue in the reporting bug. He said he can get to it this Monday the 14th.

**Comment by rlobillo** [comment]
/label qe-approved
 
 upgrade to build with candidate fix worked fine:

```
Cluster Bot
APP job job build openshift/machine-config-operator#3394 succeeded

$ oc adm upgrade --to-image=registry.build05.ci.openshift.org/ci-ln-754tblb/release@sha256:05527cad232e769653ff04a4eb706555bed055ffa410074e5d2bc8d851e2f4f3 --allow-explicit-upgrade --force
warning: The requested upgrade image is not one of the available updates.You have used --allow-explicit-upgrade for the update to proceed anyway
warning: --force overrides cluster verification of your supplied release image and waives any update precondition failures.
Updating to release image registry.build05.ci.openshift.org/ci-ln-754tblb/release@sha256:05527cad232e769653ff04a4eb706555bed055ffa410074e5d2bc8d851e2f4f3

$ oc get clusterversion
NAME      VERSION                                                   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.12.0-0.ci.test-2022-11-14-115229-ci-ln-754tblb-latest   True        False         12h     Cluster version is 4.12.0-0.ci.test-2022-11-14-115229-ci-ln-754tblb-latest
$ oc get clusteroperator/machine-config
NAME             VERSION                                                   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
machine-config   4.12.0-0.ci.test-2022-11-14-115229-ci-ln-754tblb-latest   True        False         False      14h     
$ oc get nodes
NAME                          STATUS   ROLES                  AGE   VERSION
ostest-9v8qp-master-0         Ready    control-plane,master   14h   v1.25.2+a582dda
ostest-9v8qp-master-1         Ready    control-plane,master   14h   v1.25.2+a582dda
```

**Comment by martinkennelly** [comment]
@sinnykumari Thanks for taking a look.

**Comment by sinnykumari** [comment]
Thanks for confirming these test failures.
/approve
